ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+### Return values and errors
+
+An activity can communicate its result to the workflow by returning a value or raising an error. The
+return values will get serialized as a `Payload` object similar to activity input arguments.
+Similarly, an error will get serialized and re-raised from within the workflow (if an activity
+called "synchronously").
+
+The main idea here is to make it feel (from the workflow's perspective) like a local invocation of
+an activity without any additional constructs in the way.
+
+```ruby
+class MakePaymentActivity < Temporal::Activity
+  PaymentMethodNotFound = Class.new(Temporal::Error::ActivityFailure)
+  PaymentFailed = Class.new(Temporal::Error::ActivityFailure)
+
+  def execute(payment_method_id)
+    payment_method = PaymentMethod.find(payment_method_id)
+    raise UserNotFound, "No payment_method with id #{payment_method_id}" unless payment_method
+
+    result = PaymentService.execute(user.payment_method)
+    raise PaymentFailed, "Unable to process payment: #{result.error}" if result.failure?
+
+    result.payment_id
+  end
+end
+```
+
+Notes:
+
+- Errors defined within the activity class are namespaced as `MyActivity::PaymentFailed`,
+  which is a very useful property
+  - This will however not work if a workflow does not have access to the activity definition
+  - In this case it will get raised as a generic `Temporal::Error::ActivityFailure` with an original
+    error's class name available via `#original_error`
+- Subclassing `Temporal::Error::ActivityFailure` allows us to inject extra information into the
+  error (such as activity name, id, etc)
+- All errors will also have their backtrace serialized and relayed to the caller
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
+
+The main difference is that a DataConverter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the later is for
+modifying a binary that is transmitted over the wire.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7933')
+
+config.add_data_converter(MyProtobufConverter.new)
+config.add_data_converter(MyThriftConverter) # can be a class/module based on duck-typing
+
+config.add_codec(EncryptionCodec.new(KEY_ID))
+config.add_codec(CompressionCodec)
+```
+
+A data converter is expected to have this interface:
+
+```ruby
+class Temporal::DataConverter::Base
```

Comment:
We've settled on the name `PayloadConverter` in our other (newer) SDKs

Reply:
Done — https://github.com/temporalio/proposals/pull/57/commits/804a864ebb82ec3015836942b09ea6411ad0efcd